File: compare/comparator.js
function getLocalStationSet(localResult) {
  var set = {};
  for (var i = 0; i < localResult.path.length; i++) {
    if (i === 0 || localResult.path[i].station !== localResult.path[i - 1].station) {
      set[localResult.path[i].station] = true;
    }
  }
  return set;
}

function getAmapStationSet(amapScheme) {
  var set = {};
  for (var i = 0; i < amapScheme.metro_segments.length; i++) {
    var seg = amapScheme.metro_segments[i];
    var stops = seg.all_stops || [seg.departure_stop].concat(seg.via_stops || [], [seg.arrival_stop]);
    for (var j = 0; j < stops.length; j++) {
      set[stops[j]] = true;
    }
  }
  return set;
}

function jaccardIndex(setA, setB) {
  var keysA = Object.keys(setA);
  var keysB = Object.keys(setB);
  var intersection = 0;
  for (var i = 0; i < keysA.length; i++) {
    if (setB[keysA[i]]) intersection++;
  }
  var union = keysA.length + keysB.length - intersection;
  if (union === 0) return 1.0;
  return intersection / union;
}

function compareTransferStations(localTransfers, amapTransfers) {
  var localSet = {};
  for (var i = 0; i < localTransfers.length; i++) {
    localSet[localTransfers[i]] = true;
  }
  var amapSet = {};
  for (var i = 0; i < amapTransfers.length; i++) {
    amapSet[amapTransfers[i]] = true;
  }

  var common = [];
  var localOnly = [];
  var amapOnly = [];

  var allKeys = {};
  for (var k in localSet) allKeys[k] = true;
  for (var k in amapSet) allKeys[k] = true;

  for (var k in allKeys) {
    if (localSet[k] && amapSet[k]) common.push(k);
    else if (localSet[k]) localOnly.push(k);
    else amapOnly.push(k);
  }

  return { common: common, local_only: localOnly, amap_only: amapOnly };
}

function compareSegmentLines(localSegments, amapSegments) {
  var localLines = localSegments.map(function (s) { return s.line; });
  var amapLines = amapSegments.map(function (s) { return s.line_short; });
  return { local_lines: localLines, amap_lines: amapLines, lines_match: JSON.stringify(localLines) === JSON.stringify(amapLines) };
}

function compare(localResult, amapResult, config) {
  if (localResult.error && amapResult.error) {
    return { error: 'Both failed', local_error: localResult.error, amap_error: amapResult.error };
  }

  var amapScheme = null;
  if (!amapResult.error && amapResult.schemes && amapResult.schemes.length > 0) {
    amapScheme = amapResult.schemes[0];
  }

  var result = {
    local_ok: !localResult.error,
    amap_ok: !!amapScheme,
  };

  if (localResult.error) {
    result.local_error = localResult.error;
  }
  if (amapResult.error) {
    result.amap_error = amapResult.error;
  }

  if (!result.local_ok && !result.amap_ok) return result;

  if (result.local_ok) {
    result.local_time_min = localResult.total_time + (config.WAIT_TIME_MIN || 3);
    result.local_base_time_min = localResult.total_time;
    result.local_transfers = localResult.transfers;
    result.local_station_count = localResult.station_count;
    result.local_segments = localResult.segments;
    result.local_transfer_stations = localResult.transfer_stations;
    result.local_station_set = getLocalStationSet(localResult);
  }

  if (result.amap_ok) {
    result.amap_time_min = amapScheme.duration_sec / 60;
    result.amap_walking_m = amapScheme.walking_distance_m;
    result.amap_transfers = amapScheme.metro_transfers;
    result.amap_transfer_stations = amapScheme.transfer_stations;
    result.amap_metro_segments = amapScheme.metro_segments;
    result.amap_has_non_metro = amapScheme.has_non_metro;
    result.amap_station_set = getAmapStationSet(amapScheme);

    var amapStationCount = 0;
    for (var k in result.amap_station_set) amapStationCount++;
    result.amap_station_count = amapStationCount;
  }

  if (result.local_ok && result.amap_ok) {
    result.time_diff_min = result.local_time_min - result.amap_time_min;
    result.time_diff_pct = result.amap_time_min > 0 ? (result.time_diff_min / result.amap_time_min * 100) : 0;
    result.transfer_diff = result.local_transfers - result.amap_transfers;
    result.station_diff = result.local_station_count - result.amap_station_count;
    result.walking_diff_m = 0 - result.amap_walking_m;
    result.jaccard = jaccardIndex(result.local_station_set, result.amap_station_set);
    result.transfer_comparison = compareTransferStations(localResult.transfer_stations, amapScheme.transfer_stations);
    result.line_comparison = compareSegmentLines(localResult.segments, amapScheme.metro_segments);
  }

  return result;
}

module.exports = { compare: compare };

File: compare/comparator.py
def get_local_station_set(local_result):
    station_set = set()
    last_station = None
    for node in local_result["path"]:
        if node["station"] != last_station:
            station_set.add(node["station"])
            last_station = node["station"]
    return station_set


def get_amap_station_set(amap_scheme):
    station_set = set()
    for segment in amap_scheme["metro_segments"]:
        for stop in segment.get("all_stops") or [segment["departure_stop"], *segment.get("via_stops", []), segment["arrival_stop"]]:
            station_set.add(stop)
    return station_set


def jaccard_index(set_a, set_b):
    union = set_a | set_b
    if not union:
        return 1.0
    return len(set_a & set_b) / len(union)


def compare_transfer_stations(local_transfers, amap_transfers):
    local_set = set(local_transfers)
    amap_set = set(amap_transfers)
    return {
        "common": sorted(local_set & amap_set),
        "local_only": sorted(local_set - amap_set),
        "amap_only": sorted(amap_set - local_set),
    }


def normalize_line_name(name):
    if not name:
        return name
    if name.startswith("地铁") and name.endswith("号线"):
        return name
    if name.startswith("地铁") and "号线" not in name:
        return name.replace("号", "号线", 1)
    return name


def compare_segment_lines(local_segments, amap_segments):
    local_lines = [normalize_line_name(segment["line"]) for segment in local_segments]
    amap_lines = [normalize_line_name(segment["line_short"]) for segment in amap_segments]
    return {
        "local_lines": local_lines,
        "amap_lines": amap_lines,
        "lines_match": local_lines == amap_lines,
    }


def compare(local_result, amap_result, config):
    if local_result.get("error") and amap_result.get("error"):
        return {"error": "Both failed", "local_error": local_result["error"], "amap_error": amap_result["error"]}

    amap_scheme = None
    if not amap_result.get("error") and amap_result.get("schemes"):
        amap_scheme = amap_result["schemes"][0]

    result = {
        "local_ok": not local_result.get("error"),
        "amap_ok": amap_scheme is not None,
    }

    if local_result.get("error"):
        result["local_error"] = local_result["error"]
    if amap_result.get("error"):
        result["amap_error"] = amap_result["error"]
    if not result["local_ok"] and not result["amap_ok"]:
        return result

    if result["local_ok"]:
        local_station_set = get_local_station_set(local_result)
        result["local_time_min"] = local_result["total_time"] + config.WAIT_TIME_MIN
        result["local_base_time_min"] = local_result["total_time"]
        result["local_transfers"] = local_result["transfers"]
        result["local_station_count"] = local_result["station_count"]
        result["local_segments"] = local_result.get("segments", [])
        result["local_transfer_stations"] = local_result.get("transfer_stations", [])

    if result["amap_ok"]:
        amap_station_set = get_amap_station_set(amap_scheme)
        result["amap_time_min"] = amap_scheme["duration_sec"] / 60.0
        result["amap_walking_m"] = amap_scheme["walking_distance_m"]
        result["amap_transfers"] = amap_scheme["metro_transfers"]
        result["amap_transfer_stations"] = amap_scheme["transfer_stations"]
        result["amap_metro_segments"] = amap_scheme["metro_segments"]
        result["amap_has_non_metro"] = amap_scheme["has_non_metro"]
        result["amap_station_count"] = len(amap_station_set)

    if result["local_ok"] and result["amap_ok"]:
        result["time_diff_min"] = result["local_time_min"] - result["amap_time_min"]
        result["time_diff_pct"] = (result["time_diff_min"] / result["amap_time_min"] * 100) if result["amap_time_min"] else 0
        result["transfer_diff"] = result["local_transfers"] - result["amap_transfers"]
        result["station_diff"] = result["local_station_count"] - result["amap_station_count"]
        result["walking_diff_m"] = -result["amap_walking_m"]
        result["jaccard"] = jaccard_index(local_station_set, amap_station_set)
        result["transfer_comparison"] = compare_transfer_stations(result["local_transfer_stations"], result["amap_transfer_stations"])
        result["line_comparison"] = compare_segment_lines(result["local_segments"], result["amap_metro_segments"])

    return result
